ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/TransactionLogUpdatesReader.scala

```diff
@@ -0,0 +1,274 @@
+            .toSet,
+          children = raw.exerciseChildEventIds
+            .mandatory("exercise_child_events_ids")
+            .toSeq,
+          exerciseArgument = ValueSerializer.deserializeValue(
+            Compression.Algorithm
+              .assertLookup(raw.exerciseArgumentCompression)
+              .decompress(
+                raw.exerciseArgument.mandatory("exercise_argument")
+              )
+          ),
+          exerciseResult = raw.exerciseResult.map { inputStream =>
+            ValueSerializer.deserializeValue(
+              Compression.Algorithm
+                .assertLookup(raw.exerciseResultCompression)
+                .decompress(inputStream)
+            )
+          },
+          consuming = raw.eventKind == EventKind.ConsumingExercise,
+        )
+      case EventKind.Create =>
+        val createArgument =
+          raw.createArgument.mandatory("create_argument")
+        val maybeGlobalKey =
+          raw.createKeyValue.map(
+            decompressAndDeserialize(
+              Compression.Algorithm.assertLookup(raw.createKeyCompression),
+              _,
+            )
+          )
+
+        val createArgumentDecompressed = decompressAndDeserialize(
+          Compression.Algorithm.assertLookup(raw.createArgumentCompression),
+          createArgument,
+        )
+
+        TransactionLogUpdate.CreatedEvent(
+          eventOffset = raw.offset,
+          transactionId = raw.transactionId,
+          nodeIndex = raw.nodeIndex,
+          eventSequentialId = raw.eventSequentialId,
+          eventId = raw.eventId,
+          contractId = raw.contractId,
+          ledgerEffectiveTime = raw.ledgerEffectiveTime.mandatory("ledgerEffectiveTime"),
+          templateId = raw.templateId.mandatory("template_id"),
+          commandId = raw.commandId.getOrElse(""),
+          workflowId = raw.workflowId.getOrElse(""),
+          contractKey = maybeGlobalKey,
+          treeEventWitnesses = raw.treeEventWitnesses,
+          flatEventWitnesses = raw.flatEventWitnesses,
+          createArgument = createArgumentDecompressed,
+          createSignatories = raw.createSignatories.mandatory("create_signatories").toSet,
+          createObservers = raw.createObservers.mandatory("create_observers").toSet,
+          createAgreementText = raw.createAgreementText,
+        )
+      case unknownKind =>
+        throw InvalidEventKind(unknownKind)
+    }
+
+  private def decompressAndDeserialize(algorithm: Compression.Algorithm, value: InputStream) =
+    ValueSerializer.deserializeValue(algorithm.decompress(value))
+
+  final case class FieldMissingError(field: String) extends RuntimeException {
+    override def getMessage: String = s"Missing mandatory field $field"
+  }
+
+  final case class InvalidEventKind(eventKind: Int) extends RuntimeException {
+    override def getMessage: String =
+      s"Invalid event kind: $eventKind"
+  }
+
+  private object EventKind {
+    val Create = 10
```

Comment:
There are a few classes/methods common for both `TransactionLogUpdatesReader` and `ContractStateEventsReader`, so maybe we could extract some common parts out of them into a single place. What do you think? Do you prefer to keep both readers isolated?

Reply:
I would have gone for abstraction, but this one is actually going to replace the `ContractStateEventsReader`

Reply:
Ok, thanks!